tsdb/query_engine.go

```diff
@@ -0,0 +1,1042 @@
+package tsdb
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"math"
+	"os"
+	"sort"
+	"time"
+
+	"github.com/influxdb/influxdb/influxql"
+	"github.com/influxdb/influxdb/meta"
+)
+
+const (
+	// Return an error if the user is trying to select more than this number of points in a group by statement.
+	// Most likely they specified a group by interval without time boundaries.
+	MaxGroupByPoints = 100000
+
+	// Since time is always selected, the column count when selecting only a single other value will be 2
+	SelectColumnCountWithOneValue = 2
+
+	// IgnoredChunkSize is what gets passed into Mapper.Begin for aggregate queries as they don't chunk points out
+	IgnoredChunkSize = 0
+)
+
+type Executor interface {
+	Execute(chunkSize int) <-chan *influxql.Row
+}
+
+type Planner struct {
+	MetaStore interface {
+		ShardGroupsByTimeRange(database, policy string, min, max time.Time) (a []meta.ShardGroupInfo, err error)
+		NodeID() uint64
+	}
+
+	Cluster interface {
+		NewRawMapper(shardID uint64, stmt string) (*RawMapper, error)
+		NewAggMapper(shardID uint64, stmt string) (*AggMapper, error)
+	}
+
+	store *Store
+
+	Logger *log.Logger
+}
+
+func NewPlanner(store *Store) *Planner {
+	return &Planner{
+		store:  store,
+		Logger: log.New(os.Stderr, "[planner] ", log.LstdFlags),
+	}
+}
+
+// Plan creates an execution plan for the given SelectStatement and returns an Executor.
+func (p *Planner) Plan(stmt *influxql.SelectStatement) (Executor, error) {
+	shards := map[uint64]meta.ShardInfo{} // Shards requiring mappers.
+
+	for _, src := range stmt.Sources {
+		mm, ok := src.(*influxql.Measurement)
+		if !ok {
+			return nil, fmt.Errorf("invalid source type: %#v", src)
+		}
+
+		// Replace instances of "now()" with the current time, and check the resultant times.
+		stmt.Condition = influxql.Reduce(stmt.Condition, &influxql.NowValuer{Now: time.Now().UTC()})
+		tmin, tmax := influxql.TimeRange(stmt.Condition)
+		if tmax.IsZero() {
+			tmax = time.Now()
```

Comment:
That sounds fair.  It does feel odd that we artificially cap the future on an open query, although this could be considered a characteristic of time series data.  But agree it should be in a different PR.